addons/payment_alipay/models/payment_transaction.py

```diff
@@ -183,44 +81,90 @@ def _alipay_form_get_tx_from_data(self, data):
 
         return txs
 
-    def _alipay_form_get_invalid_parameters(self, data):
-        invalid_parameters = []
+
+    def _process_feedback_data(self, data):
+        if self.provider != 'alipay':
+            return super()._process_feedback_data(data)
+        if self.state in ['done']:
+            _logger.info('Alipay: trying to validate an already validated tx (ref %s)', self.reference)
+            return True
 
         if float_compare(float(data.get('total_fee', '0.0')), (self.amount + self.fees), 2) != 0:
-            invalid_parameters.append(('total_fee', data.get('total_fee'), '%.2f' % (self.amount + self.fees)))  # mc_gross is amount + fees
+            # mc_gross is amount + fees
+            raise ValidationError(_("The paid amount does not match the total + fees."))
         if self.acquirer_id.alipay_payment_method == 'standard_checkout':
             if data.get('currency') != self.currency_id.name:
-                invalid_parameters.append(('currency', data.get('currency'), self.currency_id.name))
+                raise ValidationError(_("The currency returned by Alipay does not match the transaction currency."))
         else:
             if data.get('seller_email') != self.acquirer_id.alipay_seller_email:
-                invalid_parameters.append(('seller_email', data.get('seller_email'), self.acquirer_id.alipay_seller_email))
-        return invalid_parameters
-
-    def _alipay_form_validate(self, data):
-        if self.state in ['done']:
-            _logger.info('Alipay: trying to validate an already validated tx (ref %s)', self.reference)
-            return True
+                raise ValidationError(_("The seller email does not match the configured Alipay account."))
 
         status = data.get('trade_status')
-        res = {
-            'acquirer_reference': data.get('trade_no'),
-        }
         if status in ['TRADE_FINISHED', 'TRADE_SUCCESS']:
-            _logger.info('Validated Alipay payment for tx %s: set as done' % (self.reference))
-            date_validate = fields.Datetime.now()
-            res.update(date=date_validate)
-            self._set_transaction_done()
-            self.write(res)
-            self.execute_callback()
-            return True
+            _logger.info('Validated Alipay payment for tx %s: set as done' % self.reference)
+            self._set_done()
         elif status == 'TRADE_CLOSED':
-            _logger.info('Received notification for Alipay payment %s: set as Canceled' % (self.reference))
-            res.update(state_message=data.get('close_reason', ''))
-            self._set_transaction_cancel()
-            return self.write(res)
+            _logger.info('Received notification for Alipay payment %s: set as Canceled' % self.reference)
+            self._set_canceled()
```

Comment:
Once we receive a notification, we send 'success' message' so their system know that they can stop ping us.
If we receive   TRADE_CLOSED multiple time, the transaction is canceled on our side and that's it. They will stop notify us once we respond to their notification.
I think the _process_feedback_data is always called twice:
1) after the return from their server
2) for the first notify (2min after the transaction)

Reply:
Alrighty. I'd still check that the tx state is not already 'cancel' (same for the other states).
Otherwise, the logs will be filled with warnings (see `_update_sate` method): "tried to move from illegal source state 'cancel' to target state 'cancel', blablabla"